elastic/logstash
# geoip integrate air-gapped bootstrap script

enhancement

## Release notes

Bootstrap air-gapped environment for GeoIP database service

GeoIP plugin is expected to check update every day. For an air-gapped environment, Users can use a script, elasticsearch-geoip, to bootstrap a mock server to interact with Logstash. Set `xpack.geoip.download.endpoint` to the mock server.


## What does this PR do?

This PR integrates Elasticsearch bootstrap script to help user keep Logstah geoip plugin run without online update check.
Add `xpack.geoip.download.endpoint` option to config geoip database service endpoint. 
The default value is `https://geoip.elastic.co/v1/database`.
Users can point to `http://localhost:8080/overview.json` when using the [script](https://www.elastic.co/guide/en/elasticsearch/reference/master/geoip-processor.html#ingest-geoip-downloader-endpoint) to bootstrap nginx docker


## Why is it important/What is the impact to the user?

For security reasons, not all Logstash are free to access the internet. After 30 days without update, Maxmind EULA databases expire. The solution is setting up a mock server that mocks the interactions, a check update response and a download response.

## Checklist




## Author's Checklist


- [ ] please follow the steps to test it locally
- [ ] please check the [doc](https://github.com/logstash-plugins/logstash-filter-geoip/pull/189) if it is clear 

## How to test this PR locally



- download latest elasticsearch
- copy $ES_HOME/modules/ingest-geoip/*.mmdb to YOUR_DB_DIR
- generate mock response,  `./bin/elasticsearch-geoip -s YOUR_DB_DIR`, `overview.json` file is generated
- in `logstash.yml`, set `xpack.geoip.download.endpoint: "http://localhost:8080/overview.json"`
- start mock server nginx, `docker run -d --name=nginx_geoip -p 8080:80 -v YOUR_DB_DIR:/usr/share/nginx/html:ro nginx`
- start Logstash with geoip. You should see no error in log and see database download in $LS_HOME/data/plugins/filters/geoip
```
    input {
      heartbeat {
        message => '{"host": "55.159.212.43", "app": "barfoo", "amount": 82.95}'
        interval => 10
      }
    }
    filter {
      json {
        source => "message"
      }
      geoip {
        source => "[host]"
        target => "[server]"
      }
    }
    output {
      stdout { codec => rubydebug }
    }
```
